src/base58.cpp

```diff
@@ -87,40 +88,37 @@ static const int log256_58Ratio = ceil((std::log(256) / std::log(58)) * baseScal
     return true;
 }
 
+auto Clone(const Span<const unsigned char>& input, int start) -> std::vector<unsigned char> {
+    return std::vector<unsigned char>(input.begin() + start, input.end());
+}
+
 std::string EncodeBase58(Span<const unsigned char> input)
 {
-    // Skip & count leading zeroes.
-    int zeroes = 0;
-    int length = 0;
-    while (input.size() > 0 && input[0] == 0) {
-        input = input.subspan(1);
-        zeroes++;
+    auto leading{0U};
+    while (leading < input.size() && input[leading] == 0) {
+        ++leading;
     }
+
     int size = 1 + input.size() * log256_58Ratio / baseScale;
-    std::vector<unsigned char> b58(size);
-    // Process the bytes.
-    while (input.size() > 0) {
-        int carry = input[0];
-        int i = 0;
-        // Apply "b58 = b58 * 256 + ch".
-        for (std::vector<unsigned char>::reverse_iterator it = b58.rbegin(); (carry != 0 || i < length) && (it != b58.rend()); it++, i++) {
-            carry += 256 * (*it);
-            *it = carry % 58;
-            carry /= 58;
-        }
+    std::string result;
+    result.reserve(leading + size);
+    result.assign(leading, '1');
 
-        assert(carry == 0);
-        length = i;
-        input = input.subspan(1);
+    auto inputClone = Clone(input, leading);
+    for (auto i{0U}; i < inputClone.size();) {
+        int64_t remainder{0};
+        for (auto j{i}; j < inputClone.size(); ++j) { // Calculate next digit, dividing inputAsLongs
+            long accumulator = (remainder << 8) | inputClone[j];
+            inputClone[j] = accumulator / 58;
+            remainder = accumulator % 58;
+        }
+        result += pszBase58[remainder];
+        while (i < inputClone.size() && inputClone[i] == 0) {
+            ++i; // Skip new leading zeros
+        }
     }
-    // Translate the result into a string.
-    std::string str;
-    std::vector<unsigned char>::iterator it = b58.begin() + (size - length);
-    str.reserve(zeroes + (b58.end() - it));
-    str.assign(zeroes, '1');
-    while (it != b58.end())
-        str += pszBase58[*(it++)];
-    return str;
+    std::reverse(result.begin() + leading, result.end());
```

Comment:
we're collecting directly into the result, but need a linear reversal at the end